File: python/aide_frame/qrcode_utils.py
"""
QR code generation utilities for aide-frame applications.

Provides functions for generating QR codes as:
- Base64 data URLs (for web embedding)
- PIL Image objects (for further processing)
- Files on disk

All functions lazy-load the qrcode library to avoid import overhead
when QR codes are not used.

Usage:
    from aide_frame.qrcode_utils import generate_qr_base64, generate_qr_image

    # For web display
    data_url = generate_qr_base64("https://example.com")

    # For image processing
    img = generate_qr_image("https://example.com")
"""

import base64
import io
from typing import Optional, Tuple

from .log import logger


def is_available() -> bool:
    """Check if qrcode library is installed."""
    try:
        import qrcode
        return True
    except ImportError:
        return False


def generate_qr_image(
    data: str,
    box_size: int = 10,
    border: int = 2,
    fill_color: str = "black",
    back_color: str = "white",
):
    """
    Generate a QR code as a PIL Image.

    Args:
        data: Text or URL to encode
        box_size: Size of each QR code box in pixels
        border: Border size in boxes
        fill_color: QR code color (default black)
        back_color: Background color (default white)

    Returns:
        PIL Image object, or None if qrcode library not available

    Example:
        img = generate_qr_image("https://example.com")
        img.save("qr.png")
    """
    try:
        import qrcode
    except ImportError:
        logger.warning("qrcode library not installed. Run: pip install qrcode[pil]")
        return None

    qr = qrcode.QRCode(
        version=1,
        error_correction=qrcode.constants.ERROR_CORRECT_M,
        box_size=box_size,
        border=border,
    )
    qr.add_data(data)
    qr.make(fit=True)

    return qr.make_image(fill_color=fill_color, back_color=back_color)


def generate_qr_base64(
    data: str,
    box_size: int = 10,
    border: int = 2,
    fill_color: str = "black",
    back_color: str = "white",
) -> Optional[str]:
    """
    Generate a QR code as a base64 data URL.

    Args:
        data: Text or URL to encode
        box_size: Size of each QR code box in pixels
        border: Border size in boxes
        fill_color: QR code color
        back_color: Background color

    Returns:
        Data URL string like "data:image/png;base64,..." or None if unavailable

    Example:
        data_url = generate_qr_base64("https://example.com")
        # Use in HTML: <img src="{data_url}">
    """
    img = generate_qr_image(data, box_size, border, fill_color, back_color)
    if img is None:
        return None

    buffer = io.BytesIO()
    img.save(buffer, format='PNG')
    buffer.seek(0)
    img_base64 = base64.b64encode(buffer.getvalue()).decode('utf-8')

    return f"data:image/png;base64,{img_base64}"


def generate_qr_file(
    data: str,
    output_path: str,
    box_size: int = 10,
    border: int = 2,
    fill_color: str = "black",
    back_color: str = "white",
) -> bool:
    """
    Generate a QR code and save to file.

    Args:
        data: Text or URL to encode
        output_path: Path to save the PNG file
        box_size: Size of each QR code box in pixels
        border: Border size in boxes
        fill_color: QR code color
        back_color: Background color

    Returns:
        True if successful, False otherwise
    """
    import os

    img = generate_qr_image(data, box_size, border, fill_color, back_color)
    if img is None:
        return False

    try:
        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        img.save(output_path, "PNG")
        logger.debug(f"Generated QR code: {output_path}")
        return True
    except Exception as e:
        logger.error(f"Failed to save QR code: {e}")
        return False


def resize_qr_image(img, size: Tuple[int, int]):
    """
    Resize a QR code image (Pillow version compatible).

    Args:
        img: PIL Image object
        size: Target size as (width, height)

    Returns:
        Resized PIL Image
    """
    try:
        from PIL import Image
        resample = Image.Resampling.NEAREST
    except (ImportError, AttributeError):
        # Pillow < 9.1.0
        from PIL import Image
        resample = Image.NEAREST

    return img.resize(size, resample)


__all__ = [
    'is_available',
    'generate_qr_image',
    'generate_qr_base64',
    'generate_qr_file',
    'resize_qr_image',
]

File: js/aide_frame/src/qrcode-utils.js
/**
 * QR code generation utilities for aide-frame applications.
 *
 * Provides functions for generating QR codes as:
 * - Base64 data URLs (for web embedding)
 * - Files on disk
 *
 * Uses the 'qrcode' npm package for generation.
 *
 * Usage:
 *   const { qrcodeUtils } = require('aide-frame');
 *
 *   // For web display
 *   const dataUrl = await qrcodeUtils.generateBase64("https://example.com");
 *
 *   // Check availability
 *   if (qrcodeUtils.isAvailable()) {
 *       // QR code generation is available
 *   }
 */

const { logger } = require('./log');

let qrcode = null;

/**
 * Check if qrcode library is available.
 * @returns {boolean} True if qrcode library is installed
 */
function isAvailable() {
    try {
        if (!qrcode) {
            qrcode = require('qrcode');
        }
        return true;
    } catch (e) {
        return false;
    }
}

/**
 * Generate a QR code as a base64 data URL.
 *
 * @param {string} data - Text or URL to encode
 * @param {object} options - Options
 * @param {number} options.width - Width in pixels (default: 200)
 * @param {number} options.margin - Margin in modules (default: 2)
 * @param {string} options.color - QR code color (default: "#000000")
 * @param {string} options.background - Background color (default: "#ffffff")
 * @returns {Promise<string|null>} Data URL string or null if unavailable
 */
async function generateBase64(data, options = {}) {
    if (!isAvailable()) {
        logger.warning('qrcode library not installed. Run: npm install qrcode');
        return null;
    }

    const {
        width = 200,
        margin = 2,
        color = '#000000',
        background = '#ffffff',
    } = options;

    try {
        const dataUrl = await qrcode.toDataURL(data, {
            width,
            margin,
            color: {
                dark: color,
                light: background,
            },
        });
        return dataUrl;
    } catch (e) {
        logger.error(`Error generating QR code: ${e.message}`);
        return null;
    }
}

/**
 * Generate a QR code and save to file.
 *
 * @param {string} data - Text or URL to encode
 * @param {string} filePath - Path to save PNG file
 * @param {object} options - Options (same as generateBase64)
 * @returns {Promise<boolean>} True on success, false on failure
 */
async function generateFile(data, filePath, options = {}) {
    if (!isAvailable()) {
        logger.warning('qrcode library not installed. Run: npm install qrcode');
        return false;
    }

    const {
        width = 200,
        margin = 2,
        color = '#000000',
        background = '#ffffff',
    } = options;

    try {
        await qrcode.toFile(filePath, data, {
            width,
            margin,
            color: {
                dark: color,
                light: background,
            },
        });
        return true;
    } catch (e) {
        logger.error(`Error saving QR code to file: ${e.message}`);
        return false;
    }
}

/**
 * Generate QR code as terminal string (for CLI display).
 *
 * @param {string} data - Text or URL to encode
 * @param {object} options - Options
 * @param {boolean} options.small - Use small mode (default: true)
 * @returns {Promise<string|null>} Terminal string or null if unavailable
 */
async function generateTerminal(data, options = {}) {
    if (!isAvailable()) {
        logger.warning('qrcode library not installed. Run: npm install qrcode');
        return null;
    }

    const { small = true } = options;

    try {
        const result = await qrcode.toString(data, {
            type: 'terminal',
            small,
        });
        return result;
    } catch (e) {
        logger.error(`Error generating terminal QR code: ${e.message}`);
        return null;
    }
}

module.exports = {
    isAvailable,
    generateBase64,
    generateFile,
    generateTerminal,
};
